File: picasso-commons/src/main/java/com/hendraanggrian/picasso/commons/transformation/CropSquareTransformer.java
package com.hendraanggrian.picasso.commons.transformation;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.NonNull;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class CropSquareTransformer extends Transformer {

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, boolean recycleSource) {
        final int size = Math.min(source.getWidth(), source.getHeight());
        final Bitmap target = Bitmap.createBitmap(source,
                (source.getWidth() - size) / 2,
                (source.getHeight() - size) / 2,
                size,
                size);
        if (source != target && recycleSource)
            source.recycle();
        return target;
    }

    @NonNull
    @Override
    protected Bundle keyBundle() {
        return Bundle.EMPTY;
    }
}

File: picasso-commons/src/com/hendraanggrian/picasso/commons/transformation/CropSquareTransformer.kt
package com.hendraanggrian.picasso.commons.transformation

import android.graphics.Bitmap
import android.os.Bundle

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
internal open class CropSquareTransformer : Transformer() {

    override fun transform(source: Bitmap, shouldRecycle: Boolean): Bitmap {
        val size = Math.min(source.width, source.height)
        val target = Bitmap.createBitmap(source,
                (source.width - size) / 2,
                (source.height - size) / 2,
                size,
                size)

        if (source != target && shouldRecycle) {
            source.recycle()
        }
        return target
    }

    override fun keyBundle(): Bundle {
        val bundle = Bundle()
        bundle.putString(Transformer.KEY_NAME, "CropSquareTransformer")
        return bundle
    }
}